### PR TITLE
Get Action URL from CLI (Review)

### PR DESCRIPTION
**Pull request by dubeejw**
Adds a `--url` flag to the `action get` command that retrieves non-web action, and web action URLs.

### CODE REVIEW

**Comment by dubeejw** [comment]
PG4 490.

File: tools/cli/go-whisk-cli/commands/action.go [hunk 1]
```diff
@@ -846,38 +861,31 @@ func printActionDeleted(entityName string) {
...
     client.Namespace = qname.namespace
     fullActionName := "/" + qname.namespace + "/" + qname.entityName
 
     action, _, err := client.Actions.Get(qname.entityName)
+
     if err != nil {
         whisk.Debug(whisk.DbgError, "client.Actions.Get(%s) error: %s\n", fullActionName, err)
         whisk.Debug(whisk.DbgError, "Unable to obtain action '%s' for web action validation\n", fullActionName)
-        err = errors.New(wski18n.T("API action '{{.name}}' does not exist", map[string]interface{}{"name": fullActionName}))
+        err = errors.New(wski18n.T("API action '{{.name}}' does not exist", map[string]interface{}{"name": fullActionName})) // wrong need 404
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: @mdeuser, I don't think we should mention API in action.go.

File: tools/cli/go-whisk-cli/commands/action.go [hunk 2]
```diff
@@ -846,38 +861,31 @@ func printActionDeleted(entityName string) {
...
     } else {
+        // wrong error messages
         err = errors.New(wski18n.T("API action '{{.name}}' is not a web action. Issue 'wsk action update {{.name}} --web true' to convert the action to a web action.",
```
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: @mdeuser, same comment about the API message.

**Comment by dubeejw** [comment]
PG approved.

**Comment by dubeejw** [comment]
@mdeuser, please review.

File: tools/cli/go-whisk-cli/commands/action.go (hunk 1 above)
Comment (mdeuser) [inline tools/cli/go-whisk-cli/commands/action.go]: Good eye.  Fix already in progress via PR #2436

File: tools/cli/go-whisk-cli/commands/action.go [hunk 3]
```diff
@@ -846,38 +861,31 @@ func printActionDeleted(entityName string) {
...
             map[string]interface{}{"name": fullActionName}))
-        weVal := getValue(action.Annotations, "web-export")
-        if (weVal == nil) {
-            whisk.Debug(whisk.DbgError, "getValue(annotations, web-export) for action %s found no value\n", fullActionName)
-        } else {
-            var webExport bool
-            var ok bool
-            if webExport, ok = weVal.(bool); !ok {
-                whisk.Debug(whisk.DbgError, "web-export annotation value (%v) is not a boolean\n", weVal)
```
Comment (mdeuser) [inline tools/cli/go-whisk-cli/commands/action.go]: Replacement `WebAction` method does not have the same debug output..

File: tools/cli/go-whisk-cli/commands/action.go [hunk 4]
```diff
@@ -822,6 +825,18 @@ func printActionGetWithField(entityName string, field string, action *whisk.Acti
     printField(action, field)
 }
 
+func printActionGetWithURL(entityName string, actionURL string) {
+    fmt.Fprintf(
+        color.Output,
+        wski18n.T(
+            "{{.ok}} got action {{.name}}, displaying url\n",
```
Comment (mdeuser) [inline tools/cli/go-whisk-cli/commands/action.go]: i think using the same `{{.ok}} got action {{.name}}` as before is sufficient (i.e. no need for the `,displaying url`).  the action contents that follow adhere to whatever filters were specified on the command line - which is currently just `--url` at the moment, but could be expanded upon in the future.

File: tools/cli/go-whisk-cli/commands/action.go [hunk 5]
```diff
@@ -918,6 +926,7 @@ func init() {
     actionInvokeCmd.Flags().BoolVarP(&flags.action.result, "result", "r", false, wski18n.T("blocking invoke; show only activation result (unless there is a failure)"))
 
     actionGetCmd.Flags().BoolVarP(&flags.common.summary, "summary", "s", false, wski18n.T("summarize action details"))
+    actionGetCmd.Flags().BoolVarP(&flags.action.url, "url", "r", false, wski18n.T("get action url"))
```
Comment (mdeuser) [inline tools/cli/go-whisk-cli/commands/action.go]: How does the CLI do when both the `--url` flag and `--summary` flag are specified.  Is one ignored (if so, how does user know in doc and in the results)?  Both honored?

And if both `--url` and the `[FIELD_FILTER]` are specified together?

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 6]
```diff
@@ -695,6 +696,59 @@ class WskBasicUsageTests
             }
     }
 
+    it should "get an action URL" in withAssetCleaner(wskprops) {
+        (wp, assetHelper) =>
+            val actionName = "action name"
+            val packageName = "package name"
+            val defaultPackageName = "default"
+            val webActionName = "web action name"
+            val nonExistentActionName = "non-existence action"
```
Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: Also try an action name that has a bunch of special characters to validate the url encoding is correct

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 7]
```diff
@@ -695,6 +696,59 @@ class WskBasicUsageTests
...
+            wsk.action.get(webActionName, url = Some(true)).
+                stdout should include(webActionPath.format(wskprops.apihost, wskprops.apiversion, encodedNamespace, defaultPackageName, encodedWebActionName))
+
+            wsk.action.get(packagedAction, url = Some(true)).
+                stdout should include(packagedActionPath.format(wskprops.apihost, wskprops.apiversion, encodedNamespace, encodedPackageName, encodedActionName))
+
+            wsk.action.get(packagedWebAction, url = Some(true)).
+                stdout should include(webActionPath.format(wskprops.apihost, wskprops.apiversion, encodedNamespace, encodedPackageName, encodedWebActionName))
+
+            wsk.action.get(nonExistentActionName, url = Some(true), expectedExitCode = NOT_FOUND)
```
Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: Include a test with both `--url` and `--summary` specified.   Also with `--url` and `[FIELD_FILTER]`

File: tools/cli/go-whisk-cli/commands/action.go (hunk 5 above)
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: Updated the command usage to inform the user that `--url`, `--summary`, and field filtering do not work in conjunction with each other. I'm think we need to go through all the commands and see which flags work together and throw errors when multiple flags that incompatible with each other are specified.

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala (hunk 7 above)
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: @mdeuser, test added.

File: tools/cli/go-whisk-cli/commands/action.go (hunk 3 above)
Comment (dubeejw) [inline tools/cli/go-whisk-cli/commands/action.go]: @mdeuser, added a debug statement in the WebAction function.

File: tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala [hunk 8]
```diff
@@ -695,6 +696,63 @@ class WskBasicUsageTests
             }
     }
 
+    it should "get an action URL" in withAssetCleaner(wskprops) {
+        (wp, assetHelper) =>
+            val actionName = "action name@_-."
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/cli/test/WskBasicUsageTests.scala]: @mdeuser, these are the allowed special characters: `@ _-.`.

**Comment by dubeejw** [comment]
@mdeuser, any thing else on this one?

**Comment by mdeuser** [comment]
LGTM

**Pull request merged**